### ISSUE TITLE
client version 1.39 is too new. Maximum supported API version is 1.37 - Running v1.37 (18.03)

### ISSUE THREAD

**Issue by icornett** [terraform-providers/terraform-provider-docker#173]
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Terraform Version
Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please upgrade because your issue may have already been fixed.

Terraform v0.12.6
+ provider.docker v2.1.0

### Affected Resource(s)
Please list the resources as a list, for example:
- opc_instance
- opc_storage_volume

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

docker_container

### Terraform Configuration Files
```hcl
provider "docker" {
  host      = "tcp://127.0.0.1:2375/" # also happens with unix:///var/run/docker.sock
  version   = "~>2.0"
}

data "docker_registry_image" "ubuntu" {
  name = "ubuntu:bionic"
}

resource "docker_container" "webserver" {
  name    = "demo-websrv"
  image   = data.docker_registry_image.ubuntu.name
  restart = "unless-stopped"
  start   = true

  mounts {
    target = "/var/www"
    source = "/home/ian/demo"
    type   = "bind"
  }

  dynamic "ports" {
    for_each = var.open_ports
    content {
      internal = ports.key
      external = ports.value
    }
  }

  provisioner "local-exec" {
    command = "docker cp ${path.cwd}/scripts/setup_sshd.sh ${self.name}:/root "
  }

  provisioner "local-exec" {
    command = "docker exec -t --privileged ${self.name} ./setup_sshd.sh"
  }
}
```

variables.tf
```hcl
variable "open_ports" {
  type = map(number)
  description = "The map of internal to external ports"
  default = {
    80 = 18080,
    22 = 2200
  }
}
```

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

https://gist.github.com/icornett/c59ff07a8cd8a684f96ec62775677b62

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Expected Behavior
What should have happened?
Docker container should have been created.

### Actual Behavior
What actually happened?
Error suggesting Client API version is too new

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids
Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? Custom version of OpenStack? Tight ACLs?

Running in Azure on Ubuntu 18.04.2 LTS
```sh
ian@tfscratch:~/Terraform%20Training/terraform_101/terraform$ docker version
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Wed Jun 20 21:43:51 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Wed Jun 20 21:42:00 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

terraform.tfstate
```json
{
  "version": 4,
  "terraform_version": "0.12.6",
  "serial": 1,
  "lineage": "6a71ad40-1e05-588f-4abe-d47d2391b084",
  "outputs": {},
  "resources": [
    {
      "mode": "data",
      "type": "docker_registry_image",
      "name": "ubuntu",
      "provider": "provider.docker",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "sha256:c303f19cfe9ee92badbbbd7567bc1ca47789f79303ddcef56f77687d4744cd7a",
            "name": "ubuntu:bionic",
            "sha256_digest": "sha256:c303f19cfe9ee92badbbbd7567bc1ca47789f79303ddcef56f77687d4744cd7a"
          }
        }
      ]
    }
  ]
}
```

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

**Comment by icornett** [comment]
Oddly enough, this resolved with Docker v18.06 which uses 1.39... Very confused.

**Comment by mavogel** [comment]
Does this still occur with Docker v18.06?

**Comment by icornett** [comment]
Sorry, it appears that I mis-spoke.  It fixed in 18.09 w/ containerd.

```shell
12:49 $ docker version
Client:
 Version:           18.06.3-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        d7080c1
 Built:             Wed Feb 20 02:28:10 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.3-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       d7080c1
  Built:            Wed Feb 20 02:26:34 2019
  OS/Arch:          linux/amd64
  Experimental:     false
✔  ~/src/terraform_curriculum/terraform_101/terraform [master|✚ 1] 
12:49 $ terraform apply -auto-approve
data.docker_registry_image.ubuntu: Refreshing state...
docker_image.ubuntu: Creating...

Error: Unable to read Docker image into resource: Unable to list Docker images: Error response from daemon: client version 1.39 is too new. Maximum supported API version is 1.38

  on main.tf line 10, in resource "docker_image" "ubuntu":
  10: resource "docker_image" "ubuntu" {
```
It succeeds here:

```shell
✔  ~/src/terraform_curriculum/terraform_101/terraform [master|✚ 1] 
12:50 $ docker version
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.39 (downgraded from 1.40)
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:21:05 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.8
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       0dd43dd
  Built:            Wed Jul 17 17:07:25 2019
  OS/Arch:          linux/amd64
  Experimental:     false
✔  ~/src/terraform_curriculum/terraform_101/terraform [master|✚ 1] 
12:51 $ terraform apply -auto-approve
data.docker_registry_image.ubuntu: Refreshing state...
docker_image.ubuntu: Creating...
docker_image.ubuntu: Still creating... [10s elapsed]
docker_image.ubuntu: Creation complete after 14s [id=sha256:8106bf61615fa009d4326e677daee0b6df42d82c41d586249936ab2cde9e1c12iancornett/ubuntu-sshd]
docker_container.webserver: Creating...
docker_container.webserver: Provisioning with 'local-exec'...
```

**Comment by ibrohimislam** [comment]
> Could you verify that this error does not occur any more with Docker v18.06?

```
$ terraform init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "docker" (terraform-providers/docker) 2.1.1...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

...
```

Here's my docker version:
```
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:18:17 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:17:52 2019
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

on terraform apply:
```
$ terraform apply --target=module.api_gateway_service
module.api_gateway_server.data.alicloud_images.default: Refreshing state...
module.api_gateway_server.data.alicloud_instance_types.default: Refreshing state...
module.api_gateway_server.data.alicloud_vpcs.default: Refreshing state...
module.api_gateway_server.alicloud_security_group.api_gateway: Refreshing state... [id=sg-k1a69zsx0a3xeijnipy4]
module.api_gateway_server.alicloud_instance.api_gateway: Refreshing state... [id=i-k1a69zsx0a3xeijj0gxv]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.api_gateway_service.docker_container.apim will be created
  + resource "docker_container" "apim" {
      + attach           = false
      + bridge           = (known after apply)
      + container_logs   = (known after apply)
      + exit_code        = (known after apply)
      + gateway          = (known after apply)
      + id               = (known after apply)
      + image            = (known after apply)
      + ip_address       = (known after apply)
      + ip_prefix_length = (known after apply)
      + log_driver       = "json-file"
      + logs             = false
      + must_run         = true
      + name             = "api-manager"
      + network_data     = (known after apply)
      + restart          = "no"
      + rm               = false
      + start            = true

      + ports {
          + external = 8280
          + internal = 8280
          + ip       = "0.0.0.0"
          + protocol = "tcp"
        }
      + ports {
          + external = 8243
          + internal = 8243
          + ip       = "0.0.0.0"
          + protocol = "tcp"
        }
      + ports {
          + external = 9763
          + internal = 9763
          + ip       = "0.0.0.0"
          + protocol = "tcp"
        }
      + ports {
          + external = 9443
          + internal = 9443
          + ip       = "0.0.0.0"
          + protocol = "tcp"
        }
    }

  # module.api_gateway_service.docker_image.apim will be created
  + resource "docker_image" "apim" {
      + id     = (known after apply)
      + latest = (known after apply)
      + name   = "wso2/wso2am:2.6.0"
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.api_gateway_service.docker_image.apim: Creating...

Error: Unable to read Docker image into resource: Unable to list Docker images: Error response from daemon: client version 1.39 is too new. Maximum supported API version is 1.38

  on api_gateway_service/main.tf line 9, in resource "docker_image" "apim":
   9: resource "docker_image" "apim" {
```

**Comment by ibrohimislam** [comment]
#177 seems to solve it.

**Comment by mavogel** [comment]
Closed by #177. Will be released with `2.2.0`
